src/spinnaker.cpp

```diff
@@ -0,0 +1,1049 @@
+    }
+}
+
+void
+SpinnakerCamera::maybe_set_sample_type(SampleType target)
+{
+    CHECK(target < SampleTypeCount);
+    if (target != last_known_settings_.pixel_type) {
+        set_enum_node(camera_->PixelFormat, to_pixel_format(target));
+        last_known_settings_.pixel_type =
+          to_sample_type(*(camera_->PixelFormat));
+    }
+}
+
+void
+SpinnakerCamera::maybe_set_offset(
+  CameraProperties::camera_properties_offset_s target)
+{
+    CameraProperties::camera_properties_offset_s& last =
+      last_known_settings_.offset;
+    if (target.x != last.x) {
+        set_int_node(camera_->OffsetX, (int64_t)target.x);
+        last.x = (uint32_t)camera_->OffsetX();
+    }
+    if (target.y != last.y) {
+        set_int_node(camera_->OffsetY, (int64_t)target.y);
+        last.y = (uint32_t)camera_->OffsetY();
+    }
+}
+
+void
+SpinnakerCamera::maybe_set_shape(
+  CameraProperties::camera_properties_shape_s target)
+{
+    CameraProperties::camera_properties_shape_s& last =
+      last_known_settings_.shape;
+    if (target.x != last.x) {
+        set_int_node(camera_->Width, (int64_t)target.x);
+        last.x = (uint32_t)camera_->Width();
+    }
+    if (target.y != last.y) {
+        set_int_node(camera_->Height, (int64_t)target.y);
+        last.y = (uint32_t)camera_->Height();
+    }
+}
+
+void
+SpinnakerCamera::update_input_trigger(Trigger& trigger)
+{
+    trigger.kind = Signal_Input;
+    trigger.enable = *(camera_->TriggerMode) == genicam_on;
+    trigger.line = to_trigger_line(*(camera_->TriggerSource));
+    trigger.edge = to_trigger_edge(*(camera_->TriggerActivation));
+}
+
+void
+SpinnakerCamera::maybe_set_input_trigger_frame_start(Trigger& target)
+{
+    if (!is_equal(target, last_known_settings_.input_triggers.frame_start)) {
+        // Always disable trigger before any other configuration as in the
+        // Spinnaker Trigger.cpp example.
+        set_enum_node(camera_->TriggerMode, genicam_off);
+
+        set_enum_node(camera_->TriggerSelector, genicam_frame_start);
+        set_enum_node(camera_->TriggerSource, to_trigger_source(target.line));
+        set_enum_node(camera_->TriggerActivation,
+                      to_trigger_activation(target.edge));
+        set_enum_node(camera_->TriggerMode,
+                      target.enable ? genicam_on : genicam_off);
+
+        update_input_trigger(last_known_settings_.input_triggers.frame_start);
+    }
+}
+
+void
+SpinnakerCamera::update_output_trigger_exposure(Trigger& trigger)
+{
+    trigger.kind = Signal_Output;
+    trigger.enable = (*(camera_->LineSelector) == genicam_line_1) &&
+                     (*(camera_->LineSource) == genicam_exposure_active);
+    trigger.line = 1;
+    // TODO: check if this is the expected edge type for exposure active.
+    trigger.edge = TriggerEdge_LevelHigh;
+}
+
+void
+SpinnakerCamera::maybe_set_output_trigger_exposure(Trigger& target)
+{
+    // TODO: is there a way to disable an output line using Spinnaker?
```

Comment:
I believe there are two ways. I like the second the most.

1. Set the LineSelector to Line0, since that line doesn't support output.
2. Set the LineSource to UserOutput1. Then based on whether you want the line to be high or low at rest, set the UserOutputValue1. It might be best to leave that undefined.

Reply:
I like (2), so I did that. I only set UserOutputValue1 to false/off when this trigger is disabled. Don't have access to the camera right now, but I'm doing to push and see what happens.

Reply:
Not working yet. I'll debug this tomorrow.